hydroflow_datalog_core/src/lib.rs

```diff
@@ -214,14 +223,100 @@ fn generate_rule(
 
     let my_merge_index_lit = syn::LitInt::new(&format!("{}", my_merge_index), Span::call_site());
 
-    let after_join = match rule.rule_type {
-        RuleType::Sync(_) => {
-            parse_quote!(map(#after_join_map) -> [#my_merge_index_lit] #target_ident)
+    let after_join = if agg_keys.is_empty() {
+        let after_join_map: syn::Expr =
+            parse_quote!(|row: #flattened_tuple_type| (#(#group_by_exprs, )*));
+
+        match rule.rule_type {
+            RuleType::Sync(_) => {
+                parse_quote!(map(#after_join_map) -> [#my_merge_index_lit] #target_ident)
+            }
+            RuleType::NextTick(_) => {
+                parse_quote!(map(#after_join_map) -> next_tick() -> [#my_merge_index_lit] #target_ident)
+            }
+            RuleType::Async(_) => panic!("Async rules not yet supported"),
         }
-        RuleType::NextTick(_) => {
-            parse_quote!(map(#after_join_map) -> next_tick() -> [#my_merge_index_lit] #target_ident)
+    } else {
+        let pre_group_by_map: syn::Expr = parse_quote!(|row: #flattened_tuple_type| ((#(#group_by_exprs, )*), (#(#agg_exprs, )*)));
+
+        let only_none = group_by_exprs
+            .iter()
+            .map(|_| parse_quote!(None))
+            .collect::<Vec<syn::Expr>>();
+        let group_by_initial: syn::Expr = parse_quote!((#(#only_none, )*));
+
+        let group_by_input_types = group_by_exprs
+            .iter()
+            .map(|_| parse_quote!(_))
+            .collect::<Vec<syn::Type>>();
+        let group_by_input_type: syn::Type = parse_quote!((#(#group_by_input_types, )*));
+
+        let group_by_types = group_by_exprs
+            .iter()
+            .map(|_| parse_quote!(Option<_>))
+            .collect::<Vec<syn::Type>>();
+        let group_by_type: syn::Type = parse_quote!((#(#group_by_types, )*));
+
+        let group_by_stmts: Vec<syn::Stmt> = agg_keys
+            .iter()
+            .enumerate()
+            .map(|(i, (agg, _))| {
+                let idx = syn::Index::from(i);
+                let old_at_index: syn::Expr = parse_quote!(old.#idx);
+                let val_at_index: syn::Expr = parse_quote!(val.#idx);
+
+                let agg_expr: syn::Expr = match agg {
+                    TargetExpr::Aggregation(Aggregation { tpe, .. }) => match tpe {
+                        AggregationType::Max(_) => {
+                            parse_quote!(std::cmp::max(prev, #val_at_index))
+                        }
+                    },
+                    _ => panic!(),
+                };
+
+                parse_quote! {
+                    #old_at_index = if let Some(prev) = #old_at_index {
```

Comment:
This is fine for now, but I'm a bit concerned about the overhead due to the `Option`s. It would be nice to have a `group_by_reduce` to simplify this.

Reply:
How do you mean? What does `group_by_reduce` do?

Reply:
The initial value would always be the first element received during the tick, so then you only provide the accumulator function.

Reply:
I see